database/txn.go

```diff
@@ -0,0 +1,36 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package database
+
+import (
+	"context"
+	"database/sql"
+
+	"github.com/juju/juju/database/txn"
+)
+
+var (
+	defaultTransactionRunner = txn.NewTransactionRunner()
+)
+
+// Txn defines a generic txn function for applying transactions on a given
+// database. It expects that no individual transaction function should take
+// longer than the default timeout.
+// There are no retry semantics for running the function.
+//
+// This should not be used directly, instead the TrackedDB should be used to
+// handle transactions.
+func Txn(ctx context.Context, db *sql.DB, fn func(context.Context, *sql.Tx) error) error {
+	return defaultTransactionRunner.Txn(ctx, db, fn)
+}
+
+// Retry defines a generic retry function for applying transactions on a given
+// database. It expects that no individual transaction function should take
+// longer than the default timeout.
+//
+// This should not be used directly, instead the TrackedDB should be used to
+// handle transactions.
+func Retry(ctx context.Context, fn func() error) error {
```

Comment:
I'd love for this not to be exported if it shouldn't be called directly.

I can see the reason for it in the worker. We could make a new type to embed in that worker, if we didn't have to manipulate the worker's locks.

The locking worries me. It is quite opaque, and I've already had to adjust such a strategy due to deadlocking the worker from shutting down (we need a lock to shut down Dqlite when the loop exits).